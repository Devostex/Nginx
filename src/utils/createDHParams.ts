import asn1 from "asn1.js";
import { createDiffieHellman } from "crypto";
import { performance } from "perf_hooks";

import { logger } from "@lib/logger";
import settings from "@utils/settings";

/**
 * Creates Diffie-Hellman parameters
 * This function is synchronous but can take a long time
 * @returns {string} Diffie-Hellman Parameters
 */

export const createDHParams = (): string => {
	logger.creatingDHParams();
	const started = performance.now();
	const dhparam = asn1
		.define("", function () {
			this.seq().obj(this.key("p").int(), this.key("g").int());
		})
		.encode(
			{
				p: createDiffieHellman(settings.dhParamSize).getPrime(),
				g: 2
			},
			"pem",
			{ label: "DH PARAMETERS" }
		);

	logger.createdDHParams({ started });

	return dhparam;
};
