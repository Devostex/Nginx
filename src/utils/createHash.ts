import { createHash as hash } from "crypto";

const createHash = (from: string): string => {
	const md5 = hash("md5");

	md5.update(from, "utf8");
	return md5.digest("hex");
};
export default createHash;
