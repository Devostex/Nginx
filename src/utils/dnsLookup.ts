import { lookup } from "dns";

const dnsLookup = (hostname: string): Promise<boolean> =>
	new Promise((res) => lookup(hostname, {}, (err) => res(!err)));

export default dnsLookup;
