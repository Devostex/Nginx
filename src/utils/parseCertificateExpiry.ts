import { readFile } from "fs-extra";
import { DateTime } from "luxon";
import { pki } from "node-forge";

import log from "@utils/log";

export const parseCertificateExpiry = async (
	certificateFile: string
): Promise<DateTime | false> => {
	const certificate = await readFile(certificateFile, "");

	try {
		const cert = pki.certificateFromPem(certificate);
		return DateTime.fromJSDate(cert.validity.notAfter);
	} catch (error) {
		if (error instanceof Error)
			log.certificateParseFailed(certificateFile, error.message);
		else console.error(error);
		return false;
	}
};
