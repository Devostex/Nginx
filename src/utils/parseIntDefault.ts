import chalk from "chalk";

export const parseIntDefault = (
	string: string | undefined,
	or: number
): number => {
	if (!string) return or;

	if (isNaN(parseInt(string))) {
		console.log(
			chalk`Could not parse {dim ${string}} to a number, defaulting to ${or}`
		);
		return or;
	}
	return parseInt(string);
};
