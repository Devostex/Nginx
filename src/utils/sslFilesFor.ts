import { join } from "path";

import settings from "@utils/settings";

import { SimpleServer } from "@models/ParsedConfig";

type ServerWithNecessaryProperties = Pick<
	SimpleServer,
	"certbot_name" | "server_name"
>;

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const sslFilesFor = (server: ServerWithNecessaryProperties) => {
	const sslFile = (filename: string): string => sslFileFor(server, filename);
	return {
		ssl_certificate: sslFile("fullchain.pem"),
		ssl_certificate_key: sslFile("privkey.pem"),
		ssl_trusted_certificate: sslFile("chain.pem"),
		ssl_dhparam: settings.dhParamPath
	};
};

export const sslFileFor = (
	server: ServerWithNecessaryProperties,
	filename: string
): string =>
	join(
		settings.letsencryptPath, //						/etc/letsencrypt
		"live", //											/etc/letsencrypt/live
		server.certbot_name ?? server.server_name, //		/etc/letsencrypt/live/server
		filename //											/etc/letsencrypt/live/server/filename
	);
