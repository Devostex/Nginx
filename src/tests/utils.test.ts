import { join } from "path";

import { createDHParams } from "@utils/createDHParams";
import createHash from "@utils/createHash";
import dnsLookup from "@utils/dnsLookup";
import { parseCertificateExpiry } from "@utils/parseCertificateExpiry";
import { parseIntDefault } from "@utils/parseIntDefault";
import { sslFilesFor } from "@utils/sslFilesFor";

describe("Utilities", () => {
	test("Diffie-Hellman parameters", () => {
		const params = createDHParams();

		expect(typeof params).toBe("string");
		expect(params).toHaveLength(419);
	});

	test("Hash", () => {
		const hash = createHash("nginx");

		expect(typeof hash).toBe("string");
		expect(hash).toHaveLength(32);
		expect(hash).toBe("ee434023cf89d7dfb21f63d64f0f9d74");
	});

	test("DNS Lookup", async () => {
		expect(await dnsLookup("example.com")).toBe(true);
		expect(await dnsLookup("false.doesnotexists")).toBe(false);
	});

	test("Parse Certificate Expiry", async () => {
		const [exists, notExisting] = await Promise.all([
			parseCertificateExpiry(
				join(process.cwd(), "src", "tests", "certs", "example.pem")
			),
			parseCertificateExpiry("notExisting.pem")
		]);

		expect(exists).not.toBe(false);
		expect(notExisting).toBe(false);
		if (exists !== false)
			expect(exists.toISO()).toEqual("2013-10-04T14:47:15.000+02:00");
	});

	test("Parse Integer with Default", () => {
		const def = 10;
		expect(parseIntDefault("NaN", def)).toBe(def);
		expect(parseIntDefault("u1", def)).toBe(def);

		expect(parseIntDefault("1", def)).toBe(1);
		expect(parseIntDefault("1u", def)).toBe(1);
	});

	test("Create SSL File Paths for a Servername", () => {
		expect(sslFilesFor({ server_name: "example.com" })).toMatchSnapshot();
	});
});
