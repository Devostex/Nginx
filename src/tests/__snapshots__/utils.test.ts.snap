// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utilities Create SSL File Paths for a Servername 1`] = `
{
  "ssl_certificate": "/etc/letsencrypt/live/example.com/fullchain.pem",
  "ssl_certificate_key": "/etc/letsencrypt/live/example.com/privkey.pem",
  "ssl_dhparam": "/etc/letsencrypt/dhparams/dhparam.pem",
  "ssl_trusted_certificate": "/etc/letsencrypt/live/example.com/chain.pem",
}
`;
